File: core/policy/glob.go
// SPDX-License-Identifier: Apache-2.0

package policy

// Glob reports whether pattern matches s using the language-neutral wildcard
// syntax the config-glob classification layer speaks (design 03 §4, layer 4).
// Two metacharacters are recognised: "*" matches any run of characters
// (including the empty run and '.'), and "?" matches exactly one character.
//
// Unlike path.Match there is no separator semantics: error.type strings are
// dotted identifiers (`java.net.SocketTimeoutException`, `TIMEOUT`,
// `com.acme.ValidationException`) and '*' is meant to span the dots, so
// `java.net.*` matches `java.net.SocketTimeoutException` and
// `*.ValidationException` matches `com.acme.ValidationException`. Matching is
// case-sensitive and anchored at both ends — a pattern with no wildcards is an
// exact-equality test, which is how a bare code like `TIMEOUT` is written.
//
// The implementation is the classic two-pointer backtracking matcher: linear
// time, no allocation, and no regexp compilation. Because '*' and '?' are the
// only metacharacters, every other byte — including '.' — is a literal, so no
// escaping is needed for the dotted type names this operates on.
func Glob(pattern, s string) bool {
	// px/sx walk the pattern and subject; star/mark remember the position of the
	// last '*' and the subject offset it was tried against, so a later mismatch
	// backtracks by letting that '*' absorb one more character.
	var (
		px, sx     int
		star, mark = -1, 0
	)
	for sx < len(s) {
		switch {
		case px < len(pattern) && pattern[px] == '*':
			star, mark = px, sx
			px++ // tentatively let '*' match nothing; grow it on backtrack.
		case px < len(pattern) && (pattern[px] == '?' || pattern[px] == s[sx]):
			px++
			sx++
		case star >= 0:
			// Mismatch, but an earlier '*' can swallow one more subject byte.
			px = star + 1
			mark++
			sx = mark
		default:
			return false
		}
	}
	// Subject consumed; any trailing pattern must be all '*' to match the empty
	// remainder.
	for px < len(pattern) && pattern[px] == '*' {
		px++
	}
	return px == len(pattern)
}

File: sdk-java/worker/src/main/java/io/github/srjn45/rdq/worker/engine/Glob.java
/*
 * Copyright 2025-2026 Srajan Pathak
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.github.srjn45.rdq.worker.engine;

/**
 * Language-neutral wildcard glob matching for the config-glob classification
 * layer (layer 4, design 03 &sect;4). Two metacharacters: {@code *} matches any
 * run of characters (including dots and the empty string); {@code ?} matches
 * exactly one character. Matching is case-sensitive and anchored at both ends.
 *
 * <p>This is a direct Java port of {@code core/policy/glob.go}: the two-pointer
 * backtracking algorithm, same semantics. Both ends of the language cross-language
 * contract must agree on which error.type patterns match which strings.
 */
final class Glob {

    private Glob() {}

    /** Reports whether {@code pattern} matches {@code s}. */
    static boolean matches(String pattern, String s) {
        int px = 0, sx = 0, star = -1, mark = 0;
        while (sx < s.length()) {
            if (px < pattern.length() && pattern.charAt(px) == '*') {
                star = px;
                mark = sx;
                px++;
            } else if (px < pattern.length()
                    && (pattern.charAt(px) == '?' || pattern.charAt(px) == s.charAt(sx))) {
                px++;
                sx++;
            } else if (star >= 0) {
                px = star + 1;
                mark++;
                sx = mark;
            } else {
                return false;
            }
        }
        while (px < pattern.length() && pattern.charAt(px) == '*') {
            px++;
        }
        return px == pattern.length();
    }
}
